genienlp/ned/base.py

```diff
@@ -0,0 +1,333 @@
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import logging
+import os
+
+import ujson
+
+from ..data_utils.example import Entity
+from ..ned.ned_utils import has_overlap, is_banned, normalize_text
+
+logger = logging.getLogger(__name__)
+
+
+class AbstractEntityDisambiguator(object):
+    def __init__(self, args):
+        self.args = args
+        self.max_features_size = self.args.max_features_size
+
+        ### general attributes
+        with open(f'{self.args.database_dir}/wiki_entity_data/type_mappings/wiki/type_vocab_to_wikidataqid.json') as fin:
+            self.type_vocab_to_typeqid = ujson.load(fin)
+            self.typeqid_to_type_vocab = {v: k for k, v in self.type_vocab_to_typeqid.items()}
+        with open(f'{self.args.database_dir}/es_material/typeqid2id.json') as fin:
+            self.typeqid2id = ujson.load(fin)
+            self.id2typeqid = {v: k for k, v in self.typeqid2id.items()}
+
+        #### almond specific attributes
+        self.all_schema_types = set()
+
+        # keys are normalized types for each thingtalk property, values are a list of wiki types
+        self.almond_type_mapping = dict()
+
+        # a list of tuples: each pair includes a wiki type and their normalized type
+        self.wiki2normalized_type = list()
+
+        if self.args.almond_type_mapping_path:
+            # read mapping from user-provided file
+            with open(os.path.join(self.args.root, self.args.almond_type_mapping_path)) as fin:
+                self.almond_type_mapping = ujson.load(fin)
+            self.update_wiki2normalized_type()
+        else:
+            # this file contains mapping between normalized types and wiki types *per domain*
+            # we will choose the subset of domains we want via ned_domains
+            with open(
+                os.path.join(os.path.dirname(os.path.abspath(__file__)), 'database_files/almond_type_mapping.json')
+            ) as fin:
+                almond_type_mapping_all_domains = ujson.load(fin)
+            # only keep subset for provided domains
+            for domain in self.args.ned_domains:
+                self.almond_type_mapping.update(almond_type_mapping_all_domains[domain])
+            self.update_wiki2normalized_type()
+        ####
+
+        self.unk_id = 0
+        self.unk_type = self.id2typeqid[self.unk_id]
+        self.ckpt_name = None
+
+    def update_wiki2normalized_type(self):
+        for normalized_type, titles in self.almond_type_mapping.items():
+            for title in titles:
+                self.wiki2normalized_type.append((title, normalized_type))
+
+    def process_examples(self, examples, split_path, utterance_field):
```

Comment:
How can this method have an implementation if this is an abstract class?

Reply:
move to base class